server/lib/actors/hot_test.go

```diff
@@ -0,0 +1,250 @@
+	entries := make([]FetchResult, ENTRY_PER_FETCH)
+	i := 0
+	if len(estimates) > 0 {
+		sampleEntries := estimates[len(estimates) - 1]
+		//log.Printf("sampleEntries: %v", sampleEntries)
+		for ;i < int(float64(1 - NEWLY_ADD_RATE) * ENTRY_PER_FETCH); i++ {
+			//add score to existing entry
+			sample := sampleEntries.results[rand.Int31n(int32(len(sampleEntries.results)))]
+			up,down := int32(rand.Int31n(50) + 1), int32(rand.Int31n(50) + 1)
+			entries[i] = FetchResult {
+				id: sample.id,
+				score: up - down,
+				parent: sample.parent,
+				vote: uint32(up + down),
+			}
+		}
+	}
+	//make clock backward to assure all generated ids have timestamp before current getClock()
+	genClock := getClock().Add(-1 * time.Second)
+	for ;i < ENTRY_PER_FETCH; i++ {
+		//add new entry
+		up,down := int32(rand.Int31n(50) + 1), int32(rand.Int31n(50) + 1)
+		entries[i] = FetchResult {
+			id: proto.UUID(node.GenUUID(genClock)),
+			score: up - down,
+			parent: proto.UUID(rand.Int31n(50) + 1),
+			vote: uint32(up + down),
+		}
+		//parentMap[entries[i].id] = entries[i].parent
+	}
+	log.Printf("testInmemVoteFetcher: %v", entries)
+	estimates = append(estimates, fetchCache {
+		fetchAt: getClock(),
+		results: entries,
+	})
+	return entries, nil	
+}
+
+func filterSummery(btyp proto.TopicListRequest_BucketType, b *hotbucket) *hotbucket {
+	if btyp == proto.TopicListRequest_Flame {
+		return b.filter(func (e *HotEntry) bool {
+			return e.flamy()
+		}).sort(btyp).truncate(ALLTIME_SUMMERY_LIMIT)
+	} else {
+		return b.sort(btyp).truncate(ALLTIME_SUMMERY_LIMIT)
+	}
+}
+
+func testInmemVoteParentFetcher(start, end time.Time, btyp proto.TopicListRequest_BucketType) ([]FetchResult, error) {
+	tmp := NewHotBucket(ENTRY_PER_FETCH)
+	for _, e := range estimates {
+		if start.Unix() <= e.fetchAt.Unix() && e.fetchAt.Unix() <= end.Unix() {
+			tmp.addResults(e.results)
+		} else if start.Unix() <= 0 {
+			log.Printf("testInmemVoteParentFetcher: alltime query but missing: %v %v", e.fetchAt, end)
+		}
+	}
+	tmp = filterSummery(btyp, tmp)
+	ret := make([]FetchResult, len(tmp.total.entries))
+	for i, e := range tmp.total.entries {
+		ret[i] = FetchResult{ id: e.Id, parent: tmp.idParentMap[e.Id], vote: e.Vote, score: e.Score }
+	}
+	lastFilter = end
+	log.Printf("testInmemVoteParentFetcher: %v %v %v", btyp, lastFilter, ret)
+	return ret, nil
+}
+
+func testInmemFetcher(btyp proto.TopicListRequest_BucketType, start, end time.Time, locale string) ([]FetchResult, error) {
+	switch btyp {
+	case proto.TopicListRequest_Rising:
+		return testInmemVoteFetcher()
+	case proto.TopicListRequest_Hot, proto.TopicListRequest_Flame:
+		return testInmemVoteParentFetcher(start, end, btyp)
+	default:
+		return nil, fmt.Errorf("invalid fetchType: %v", btyp)
+	}
+}
+
+func testInmemPersister(at time.Time, s *hotbucketStore) error {
+	return nil
+}
+
+func estimateResults(btyp proto.TopicListRequest_BucketType, typ proto.TopicListRequest_QueryType, start, count uint64) []HotEntry {
```

Comment:
__[gometalinter]__
* [gocyclo] cyclomatic complexity 16 of function estimateResults() is high (> 10)